README
/* simplewebkit
   README
   Copyright (C) 2007 Free Software Foundation, Inc.

   Author: Dr. H. Nikolaus Schaller

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with this library; see the file COPYING.LIB.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

A1. INSTALLATION on GNUstep

(./configure)
./make


A2. INSTALLATION on Mac OS X with Xcode 2.5 or later

1. load file tree
2. open SimpleWebKit.xcodeproj

There are 3 targets:
a) SWK Browser with full WebKit
tests SWK Browser by using Apple WebKit (and AppKit, Foundation)

b) SWK Browser
replaces Apple WebKit by SimpleWebKit (but still uses Apple AppKit, Foundation)
and crosscompiles SWK Browser (using mySTEP.make) to run on QuantumSTEP

c) SimpleWebKit
crosscompiles (using mySTEP.make) SimpleWebKit.framework to run on QuantumSTEP


B. Short Description how this all Works Together

1. the WebView
* is the master view object and there is only one per browser (or browser tab)
* it holds the mainFrame which represents either the normal <body> or the top level <frame> or <frameset>
* if there is a <frameset> hierarchy, there are additional child WebFrames

2. the WebFrame
* is repsonsible for loading and rendering content from a specific URL
* it uses a WebDataSource to trigger loading and get callbacks
* it is also the owner of the DOMDocument tree
* JavaScript statements are evaluated in a frame context
* it is also the target of user clicks on links since it knows the base URL (through the WebDataSource)

3. the WebDataSource
* is responsible for loading data from an URL
* it may cache data and handle/synchronize loading fo subresources (e.g. for an embedded <img> tag)
* it translates the request and the response URLs
* it provides an estimated content length (for a progress indicator) and the MIMEType of the incoming data stream
* as soon as the header comes in a WebDocomentRepresentation is created and incoming segments are notified
* it also collects the incoming data, so that a WebDocomentRepresentation can handle either segments or the collected data

4. the WebDocumentRepresentation(s)
* there is one for each MIME type (the WebView provides a mapping database)
* it is responsible for parsing the incoming data stream (either completely when finished, or partially)
* and provide a better suitable representation, e.g. an NSImage or a DOMHTMLTree
* finally, it creates a WebDocumentView as the child of the WebView and attaches it to the WebFrame as the -webFrameView
* so, if you want to handle an additional MIME type, write a class that conforms to the WebDocumentRepresentation protocol

5. the DOMHTMLTree
* is only for HTML content
* is (re)built each time a new segment of HTML data comes in
* any change in the DOMHTMLTree is notified to the WebDocumentView (or one of its subviews) by setNeedsLayout

6. the WebDocumentView(s) an its subviews
* are responsible for displaying the contents of its WebDataRepresentation
* either HTML, Images, PDF or whatever (e.g. SVG, XML, ...)
* they gets notified about changes either by updates of the WebDataSource (-dadaSourceUpdated:) or directly (-setNeedsLayout:)
* if one needs layout, it must go to the DOM Tree to find out what has changed and update its size, content, children, layout etc.
* this is a little tricky/risky since the -layout method is called within -drawRect: - so changing e.g. the View frame is very
  critical and may result in drawing glitches
* for HTML, we do a simple trick: the WebDocumentView is an NSTextView and the DOMHTMLTree objects can be traversed to
  return an attributedString with embedded Tables and NSTextAttachments

7. the JavaScript engine
* is programmed according to the specificaion of ECMA-262 (http://www.ecma-international.org/publications/standards/Ecma-262.htm)
* uses a simple recursive stateless parser (could be optimized in stack useage and speed by a state-table driven approach)
* parses the script into a Tree representation in a first step
* then, evaluates the expressions and statements according to the current environement
* this allows to store scripts in translated form and reevaluate them when needed (e.g. on mouse events)
* uses Foundation for basic types (string, number, boolean, null)
* uses WebScriptObject as the base Object representation
* DOMObjects are a subclass of WebScriptObjects and therefore provide automatic bridging, so that changing a DOMHTML tree element through
  JavaScript automativally triggers the appropriate WebDocumentView notification

8. CSS
* is handled similarily to JavaScript and HTML
* i.e. loading is done automatically
* creates a DOMCSS tree (linked to document.styleSheets)
* this forms a searchable database for style definitions
* DOMHTML searches this for every tag when processing the style for the NSAttributedString


ChangeLog
2020-05-29 Riccardo Mottola <rm@gnu.org>

	* Sources/WebHTMLDocumentView.m
	Update methods signatures with characterIndex from unsigned to NSUInteger.

2017-08-20 Riccardo Mottola <rm@gnu.org>

	* Sources/DOMCSS.m
	Escape relative part before forming URL.

2017-08-02 Riccardo Mottola <rm@gnu.org>

	* Sources/WebHTMLDocumentView.m
	store "auto" and "zero" with using -1 and 0 instead of a microscopic quantity.

2014-11-04 Riccardo Mottola <rm@gnu.org>

	* Sources/WebHTMLDocumentView.m
	By Fred: Enable underlining and strikethrough for GNUstep, provide better fallback for old MacOS.

2014-11-04 Riccardo Mottola <rm@gnu.org>

	* Sources/DOMCSS.m
	* Sources/WebHTMLDocumentView.m
	Descend returns negative values, fix accordingly.

2013-10-25..28  Nikolaus Schaller <hns@computer.org>

	* fixed issues with CSS inherit and initial
	* improved translation of CSS styles to string attributes and fixed many smaller bugs
	* implemented a little "height" and "width"
	* made display:list and <li>, <ul>, <ol>, <dl> work for standard cases (HTML4)
	* reduced number of compiler warnings
	* fixed positions of form buttons and cells

2012-08-28  Wolfgang Lux  <wolfgang.lux@gmail.com>

	* GNUmakefile:
	* Sources/GNUmakefile: Revert error made by Riccardo while
	changing the makefiles.

2012-08-27 18:59-EDT Gregory John Casamento <greg.casamento@gmail.com>

	* Sources/WebHTMLDocumentView.m: Same fix as below.

2012-08-27 18:53-EDT Gregory John Casamento <greg.casamento@gmail.com>

	* Sources/DOMCSS.m: Fix double include problem. #import was including
	the DOMHTML.h file twice since it was both in the current directory
	and in the derived_src directory.

2012-08-**  Nikolaus Schaller <hns@computer.org>

	* reworked CSS processing to provide computedStyleForElement:pseudoElement:
	* reworked spliceTo: to translate CSS to NSAttributedString attributes
	* added default.css to define HTML4 -> CSS translation
	* removed (almost) all direct translation of attributes to style (attributes remain used for forms and table layout)
	* fixed parsing of tag[attribute=value] style
	* fixed white-space merging rules so that normal, pre and nowrap are working

2012-01-19 Riccardo Mottola <rm@gnu.org>

	* Sources/DOMHTML.m
	Revert.

2012-01-18 Riccardo Mottola <rm@gnu.org>

	* Sources/DOMHTML.m
	Really copy value to cell.

2011-05-16  Nikolaus Schaller <hns@computer.org>
	* fixed issue in scanning <script>...</script>
	* fixed signature of - (NSView <WebDocumentView> *) documentView;
	* checked with Clang for potential issues and fixed two potential leaks
	* optimized CSS parser a little

2011-04-14  Nikolaus Schaller <hns@computer.org>
	* fixed bug in DOMCSSValueList which resulted in delayed segfault in NSArrayEnumerator
	* minor cosmetics

2010-10-24 Riccardo Mottola <rmottola@users.sf.net>

	* Sources/DOMCSS.m
	Fixed some c99-isms

2010-10-23  Nikolaus Schaller <hns@computer.org>

	* fixed issue with <noframes> (which was counted as <frame> when determining the number of rows/columns of a <frameset>)
	* fixed problem with processing CSS values as strings
	* made a workaround for a problem in @import() CSS rules (needs to be fixed differently)

2010-10-22 Riccardo Mottola <rmottola@users.sf.net>

	* Source/DOMHTML.m
	Enable setting of ToolTips with the link url.

2010-10-03 Riccardo Mottola <rmottola@users.sf.net>

	*  Sources/DOMCore.h
	Enable renaming only if GCC has major version = 2

	* Sources/WebPlugin.h
	* Sources/WebPluginContainer.h
	Do not include Cocoa.h but directly Foundation and the necessary headers.

2010-08-03++  Nikolaus Schaller <hns@computer.org>
	* fixed some of the Issues reported on http://projects.goldelico.com/p/swk/issues/ (see commit messages)
	* worked on CSS parsing to be more complete for correct CSS files
	* worked on CSS stylesheet database search (match rules with nodes)
	* apply color: style
	* apply display: none

2010-08-03 Riccardo Mottola <rmottola@users.sf.net>

	* Sources/WebFrameView.m:
	Enable autohiding for GNUstep.
	
2010-05-20 00:07-EDT Gregory John Casamento <greg.casamento@gmail.com>

	* Palette/GNUmakefile:
	* Palette/GNUmakefile.preamble: Makefiles.
	* Palette/palette.table: Defines the classes to be imported.
	* Palette/SWKPalette.gorm: Interface file.
	* Palette/SWKPalette.[hm]: Implementation of the palette.
	* Palette/SWKPalette.tiff: Image.
	* Sources/WebView.m: Added an implementation of initWithCoder:.

2009-08-12++  Nikolaus Schaller <hns@computer.org>
	* Sources/NSXMLParser.m: fixed a problem with entities like &#9660;
	* Sources/WebDataSource.m
	  - fixed problem with committing when no response arrived (i.e. no network available)
		- fixed problem with relative and absolute URLs for subresources (did not notice that they are the same)
		- fixed problem where a completed subresource could call -finishedLoadingWithDataSource: twice (leading to load errors)
	* Sources/DOMCSS.*: added most CSS functionality
		- worked on class hierarchy to be more compatible to WebKit
		- parse CSS3 documents into DOM CSS tree (incl. correct handling and loading of @import)
		- parse <style> content and load and parse subresources <link rel="stylesheet">;
	* SWK Browser
		- fixed a bug in the DOM Hierarchy inspector (did not retain tree nodes)
		- fixed deleting resources from the Activity Window if pages are reloaded
		- added a CSS DOM Hierarchy inspector to the Debug menu

2009-07-21 Riccardo Mottola <rmottola@users.sf.net>
	* Sources/WebPreferences.m: return system font families
	* Sources/DOMHTML.m: use system font if font not found
	* Sources/WebHTMLDocumentView.m: draw HR better

2009-03-31  Nikolaus Schaller <hns@computer.org>
	* fixed correct translation and display of <frameset> and <frame>
	* fixed issue with dangling parent pointer in DOMNodes
	* added a View Hierarchy debug window to SWK Browser

2009-03-21 Riccardo Mottola <rmottola@users.sf.net>
	* Sources/NSXMLParser: change from end pointer to vp.

2008-07-12  Nikolaus Schaller <hns@computer.org>

Sources/DOMCSS.h
/* simplewebkit
   DOMCSS.h

   Copyright (C) 2007 Free Software Foundation, Inc.

   Author: Dr. H. Nikolaus Schaller

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with this library; see the file COPYING.LIB.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

#import <WebKit/DOMCore.h>

@class DOMStyleSheet;

@interface DOMStyleSheetList : DOMObject
{
	NSArray *items;	// of type DOMCSSStyleSheet
}

- (void) _addStyleSheet:(DOMStyleSheet *) sheet;
- (unsigned) length;
- (DOMStyleSheet *) item:(unsigned) index;

@end

@class DOMMediaList;

@interface DOMStyleSheet : DOMObject
{
	NSString *type;
	DOMNode *ownerNode;	// should be <style> or <link>
	DOMStyleSheet *parentStyleSheet;	// if nested (@import)
	NSString *href;
	NSString *title;
	DOMMediaList *media;	// for which media to apply
	BOOL disabled;
}

- (BOOL) disabled;
- (NSString *) href;
- (DOMMediaList *) media;
- (DOMNode *) ownerNode;
- (DOMStyleSheet *) parentStyleSheet;
- (NSString *) title;

// which of these are really public setters???
- (void) setHref:(NSString *) href;		// required to locate @import urls
- (void) setTitle:(NSString *) title;
- (void) setOwnerNode:(DOMNode *) node;	// required to load subresources (@import)

@end

@class DOMCSSRule;
@class DOMCSSRuleList;

@interface DOMCSSStyleSheet : DOMStyleSheet
{
	DOMCSSRule *ownerRule;
	DOMCSSRuleList *cssRules;
	BOOL enabled;
}

- (void) _setCssText:(NSString *) str;	// parse all rules...
- (DOMCSSRule *) ownerRule;
- (DOMCSSRuleList *) cssRules;
- (unsigned) insertRule:(NSString *) rule index:(unsigned) index;	// parse rule
- (void) deleteRule:(unsigned) index;

@end

@interface DOMMediaList : DOMObject
{
	NSString *mediatext;
	NSArray *items;	// of type NSString
}

- (NSString *) mediaText;
- (void) setMediaText:(NSString *) text;
- (unsigned) length;
- (NSString *) item:(unsigned) index;
- (void) deleteMedium:(NSString *) oldMedium;
- (void) appendMedium:(NSString *) newMedium;

@end

@class DOMCSSRule;

@interface DOMCSSRuleList : DOMObject
{
	NSArray *items;
}

- (unsigned) length;
- (DOMCSSRule *) item:(unsigned) index;

@end

enum DOMCSSRuleType
{
	DOM_UNKNOWN_RULE = 0,
	DOM_STYLE_RULE = 1,
	DOM_CHARSET_RULE = 2,
	DOM_IMPORT_RULE = 3,
	DOM_MEDIA_RULE = 4,
	DOM_FONT_FACE_RULE = 5,
	DOM_PAGE_RULE = 6
};

@interface DOMCSSRule : DOMObject
{
	DOMCSSStyleSheet *parentStyleSheet;
	DOMCSSRule *parentRule;
}

- (BOOL) _ruleMatchesElement:(DOMElement *) element pseudoElement:(NSString *) pseudoElement;
- (unsigned short) type;
- (NSString *) cssText;
- (DOMCSSStyleSheet *) parentStyleSheet;
- (DOMCSSRule *) parentRule;

@end

@interface DOMCSSCharsetRule : DOMCSSRule
{
	NSString *encoding;
}
- (NSString *) encoding;
@end

@class DOMCSSStyleDeclaration;

@interface DOMCSSFontFaceRule : DOMCSSRule
{
	DOMCSSStyleDeclaration *style;
}
- (DOMCSSStyleDeclaration *) style;
@end

@interface DOMCSSImportRule : DOMCSSRule <WebDocumentRepresentation>
{
	NSString *href;
	DOMMediaList *media;
	DOMCSSStyleSheet *styleSheet;	// the loaded style sheet
}
// FIXME: do we need this or can we use KVC as getters&setters?
- (NSString *) href;
- (DOMMediaList *) media;
- (DOMCSSStyleSheet *) styleSheet;
@end

@class DOMCSSRuleList;

@interface DOMCSSMediaRule : DOMCSSRule
{
	DOMMediaList *media;
	DOMCSSRuleList *cssRules;
}

- (DOMCSSRuleList *) cssRules;
- (unsigned) insertRule:(NSString *) rule index:(unsigned) index;
- (void) deleteRule:(unsigned) index;

@end

@interface DOMCSSPageRule : DOMCSSRule
{ // page style rule
	NSArray *selector;	// all selectors begin with :
	DOMCSSStyleDeclaration *style;
}

- (NSString *) selectorText;
- (void) setSelectorText:(NSString *) rule;
- (DOMCSSStyleDeclaration *) style;
- (void) setStyle:(DOMCSSStyleDeclaration *) s;

@end

@class DOMCSSStyleRuleSelector;

@interface DOMCSSStyleRule : DOMCSSRule
{ // standard style rule
	DOMCSSStyleRuleSelector *selector;
	DOMCSSStyleDeclaration *style;
}

- (NSString *) selectorText;	// tag class patterns
- (void) setSelectorText:(NSString *) rule;
- (DOMCSSStyleDeclaration *) style;
- (void) setStyle:(DOMCSSStyleDeclaration *) s;

@end

@interface DOMCSSUnknownRule : DOMCSSRule
@end

@class DOMCSSValue;

@interface DOMCSSStyleDeclaration : DOMObject
{
	NSMutableDictionary *items;
	NSMutableDictionary *priorities;
	DOMCSSRule *parentRule;
}

- (id) initWithString:(NSString *) style;
- (NSString *) cssText;
- (void) setCssText:(NSString *) style;
- (DOMCSSRule *) parentRule;
- (NSString *) getPropertyValue:(NSString *) propertyName;
- (DOMCSSValue *) getPropertyCSSValue:(NSString *) propertyName;
- (NSString *) removeProperty:(NSString *) propertyName;
- (NSString *) getPropertyPriority:(NSString *) propertyName;
- (void) setProperty:(NSString *) propertyName value:(NSString *) value priority:(NSString *) priority;
- (unsigned) length;
- (NSString *) item:(unsigned) index;
- (NSString *) getPropertyShorthand:(NSString *) propertyName;
- (BOOL) isPropertyImplicit:(NSString *) propertyName;

- (NSDictionary *) _items;

@end

enum DOMCSSValueType
{
	DOM_CSS_INHERIT = 0,
	DOM_CSS_PRIMITIVE_VALUE = 1,
	DOM_CSS_VALUE_LIST = 2,
	DOM_CSS_CUSTOM = 3
};

@interface DOMCSSValue : DOMObject
{
	unsigned short cssValueType;
}

- (NSString *) cssText;
- (unsigned short) cssValueType;

@end

@interface DOMCSSValueList : DOMCSSValue
{
	NSMutableArray *values;
}
- (unsigned) length;
- (DOMCSSValue *) item:(unsigned) index;
@end

enum DOMCSSPrimitiveValueType
{
	DOM_CSS_UNKNOWN = 0,
	DOM_CSS_NUMBER = 1,
	DOM_CSS_PERCENTAGE = 2,
	DOM_CSS_EMS = 3,
	DOM_CSS_EXS = 4,
	DOM_CSS_PX = 5,
	DOM_CSS_CM = 6,
	DOM_CSS_MM = 7,
	DOM_CSS_IN = 8,
	DOM_CSS_PT = 9,
	DOM_CSS_PC = 10,
	DOM_CSS_DEG = 11,
	DOM_CSS_RAD = 12,
	DOM_CSS_GRAD = 13,
	DOM_CSS_MS = 14,
	DOM_CSS_S = 15,
	DOM_CSS_HZ = 16,
	DOM_CSS_KHZ = 17,
	DOM_CSS_DIMENSION = 18,
	DOM_CSS_STRING = 19,
	DOM_CSS_URI = 20,	// url(string)
	DOM_CSS_IDENT = 21,
	DOM_CSS_ATTR = 22,	// attr(name)
	DOM_CSS_COUNTER = 23,
	DOM_CSS_RECT = 24,
	DOM_CSS_RGBCOLOR = 25,	// rgb(red, green, blue)
	/* CSS3 */
	DOM_CSS_RGBACOLOR = 26,
	DOM_CSS_HSVCOLOR = 27,
	DOM_CSS_HSVACOLOR = 28,
	DOM_CSS_CALC = 29,	// calc(formula)
	DOM_CSS_TURNS = 30,
};

@interface DOMCSSPrimitiveValue : DOMCSSValue
{
	unsigned short primitiveType;
	id value;	// NSNumber, NSString, NSValue, NSURL, NSColor, DOMCSSValue, ...
}

- (unsigned short) primitiveType;
- (void) setFloatValue:(unsigned short) unitType floatValue:(float) floatValue;
- (float) getFloatValue:(unsigned short) unitType;
- (void) setStringValue:(unsigned short) stringType stringValue:(NSString *) stringValue;
- (NSString *) getStringValue;
//- (DOMCounter *) getCounterValue;
//- (DOMRect *) getRectValue;
//- (DOMRGBColor *) getRGBColorValue;

@end

@interface DOMCSSPrimitiveValue (Private)

// evaluate relative (and absoulte) values to unitType:
- (float) getFloatValue:(unsigned short) unitType relativeTo100Percent:(float) base andFont:(NSFont *) font;

@end
